src/algorithm/geo/bounding_rect.rs

```diff
@@ -29,6 +32,33 @@ pub trait BoundingRect {
     fn bounding_rect(&self) -> RectArray;
 }
 
+// TODO: just an example to show how to impl algorithms for dyn GeometryArrayTrait
+impl BoundingRect for dyn GeometryArrayTrait {
+    fn bounding_rect(&self) -> RectArray {
+        match self.data_type() {
+            GeoDataType::Point(_) => {
+                let array = as_point_array(self);
+                let output_geoms: Vec<Option<Rect>> = array
+                    .iter_geo()
+                    .map(|maybe_g| maybe_g.map(|geom| geom.bounding_rect()))
+                    .collect();
+
+                RectArray::from(output_geoms)
+            }
+            GeoDataType::LineString(_) => {
+                let array = as_line_string_array::<i32>(self);
+                let output_geoms: Vec<Option<Rect>> = array
+                    .iter_geo()
+                    .map(|maybe_g| maybe_g.and_then(|geom| geom.bounding_rect()))
+                    .collect();
+
+                RectArray::from(output_geoms)
+            }
+            _ => unimplemented!()
+        }
+    }
+}
```

Comment:
Ideally we'll be able to use a similar macro as we already have here: https://github.com/geoarrow/geoarrow-rs/blob/906b5d5e507c918a1aa5fab2035f2a43cf155617/src/algorithm/geo/utils.rs#L8-L47

instead of copying the original function content for every branch

Reply:
I started scratch work of this in https://github.com/geoarrow/geoarrow-rs/pull/326